Hide email address in Sends

eliykat opened the pull request:
# Objective
Allow a user to hide their email address when creating a Send. Per [this feature request](https://community.bitwarden.com/t/option-to-hide-email-reference-from-the-ui/16797).

Basic spec:
* Sender has a new option when adding or editing a Send: "Hide email address".
* If "Hide email address" is enabled, the email address is not shown when viewing a Send.
* The Disable Send policy is expanded to allow an organization to disable anonymous sends only.
* Display a large warning when viewing an anonymous Send.

## Code changes

Brief overview of code changes across all repos:

* Send
  * Add new column to dbo.Send, `HideEmail`. Populate with default value of `0` in migration scripts.
  * Add new property to all Send models, `HideEmail`
  * Add new checkbox to Hide email address when adding/editing a Send
  * If email address is `null` when viewing a Send, display warning banner instead
* Policy
  * Add new options to Disable Send policy: `Disable all Sends` or `Disable anonymous Sends only`.
  * These are represented in an enum in `policy.data`
  * Update enforcement for `Disable all Sends` - we now have to check the enum value, not just whether the policy is enabled. (Lots of null checking around this to make it consistent with current policies that do not have the enum set)
  * Add enforcement for `Disable anonymous Sends`:
    *  disable the option in the create Send screen
    * disable editing of any existing Sends that have Hide Email address enabled
    * check server side as well

## To do

* Docs will need to be updated - the Receive Send page should be updated to explain the risk of anonymous Sends more fully
* UI for other clients
* Consider report abuse mechanism - for now have just done the banner to keep it simple

## Related PRs
* jslib: https://github.com/bitwarden/jslib/pull/310
* server: https://github.com/bitwarden/server/pull/1234
* desktop: https://github.com/bitwarden/desktop/pull/812
* browser: https://github.com/bitwarden/browser/pull/1740
* mobile: https://github.com/bitwarden/mobile/pull/1340

## Screenshots (do not reflect latest changes)

Creating/editing a Send:

![Screen Shot 2021-03-23 at 4 20 34 pm](https://user-images.githubusercontent.com/31796059/112102492-12493200-8bf4-11eb-897f-542568e44510.png)

![Screen Shot 2021-03-23 at 4 21 45 pm](https://user-images.githubusercontent.com/31796059/112102560-2e4cd380-8bf4-11eb-9ef5-0770f712e07a.png)

Viewing a Send:

![Screen Shot 2021-03-23 at 4 18 52 pm](https://user-images.githubusercontent.com/31796059/112102580-3442b480-8bf4-11eb-9fa8-f061f16d9b77.png)

Editing Disable Send policy:

![Screen Shot 2021-03-23 at 4 22 51 pm](https://user-images.githubusercontent.com/31796059/112102604-3dcc1c80-8bf4-11eb-9bf7-a357ab40015b.png)

vincentsalucci commented:
## Create/Edit Send
- Policy Enabled: Our usual pattern for disabled options is to show a banner at the top of the contextual area (this example: modal). I don't feel strongly about this, but just wanted to bring it up for consistency's sake.

## Access Send
- Banner: We generally don't center the content of an `app-callout` - it feels, odd, to me. Again, not something I feel strongly against, just thought I'd bring it up.

## Edit Policy
- Checkbox/Banner: What's the reason for switching the checkbox/banner order? I feel like the user should read that before deciding whether or not to enable the policy. 
- Wording: I had to read this dialog several times before understanding what the new options meant. Maybe we take the approach of radio options with some helper text for each (ex: Vault Timeout Action in Settings) and/or adjust the header.

cscharf commented:
> * Banner: We generally don't center the content of an `app-callout` - it feels, odd, to me. Again, not something I feel strongly against, just thought I'd bring it up.

Agreed. Let's ensure the text in the banner is left-aligned / paragraph format for consistency and readability (larger blocks of centered text is a little harder to read/follow.

The other feedback will likely be moot considering the style of policy will change from server+up.

eliykat commented:
* Create/Edit Send - I added a banner yesterday and removed the inline italic text, because I had the same thought. I played a bit more with the UI text today and here's how it looks now:

![Screen Shot 2021-03-25 at 8 26 11 am](https://user-images.githubusercontent.com/31796059/112391554-d1603300-8d43-11eb-82af-4a7920ccd5cc.png)

This is designed to add more list items as we expand the Send Options Policy. (Although I've kept the view model/templating logic simple for now, rather than overdoing it for options that don't exist yet)

* Access Send - I thought it looked strange but wasn't sure why - here's it after the fix:

![Screen Shot 2021-03-25 at 8 30 58 am](https://user-images.githubusercontent.com/31796059/112391988-77ac3880-8d44-11eb-9f21-4235bbb994e4.png)

I didn't think it was long enough to split into paragraphs.

* Editing policy

This has changed since we moved this into its own policy - everything feels much clearer this way, but still open to feedback on the wording:

![Screen Shot 2021-03-25 at 9 05 47 am](https://user-images.githubusercontent.com/31796059/112394986-4f730880-8d49-11eb-9e5c-dcb3f82c87b8.png)

The banner is below the Enabled checkbox as a side effect of how this template is structured. (Policy-specific controls, including the banner, are either all above or all below the Enabled checkbox.) I'll rework this, as currently each policy has its own warning banner to this effect, even though they're all identical or near-identical.

cscharf commented on src/app/organizations/manage/policy-edit.component.ts:
```diff
@@ -60,6 +60,10 @@ export class PolicyEditComponent implements OnInit {
     passGenIncludeNumber?: boolean;
 
+    // Send options
+
```
Remove this extra line below the comment

cscharf commented on src/locales/en/messages.json:
```diff
@@ -3635,6 +3638,26 @@
+    "message": "Organization users that can manage the organization's policies are exempt from this policy's enforcement."
+  },
+  "disableHideEmail": {
+    "message": "Do not allow users to hide their email address when creating or editing a Send.",
```
```suggestion
    "message": "Do not allow users to hide their email address from recipients when creating or editing a Send.",
```

cscharf commented on src/locales/en/messages.json:
```diff
@@ -3635,6 +3638,26 @@
+    "message": "The following organization policies are currently in effect:"
+  },
+  "sendDisableHideEmailInEffect": {
+    "message": "Users are not allowed to hide their email address when creating or editing a Send.",
```
```suggestion
    "message": "Users are not allowed to hide their email address from recipients when creating or editing a Send.",
```

cscharf commented on src/locales/en/messages.json:
```diff
@@ -3775,6 +3798,9 @@
     }
   },
+  "viewSendHiddenEmailWarning": {
+    "message": "The Bitwarden user who created this Send has chosen to hide their email address. You should not trust this Send or download any file unless you trust the user who created it."
```
```suggestion
    "message": "The Bitwarden user who created this Send has chosen to hide their email address. You should ensure you trust the source of this link before using or downloading its content.",
    "description": "'Send' is a noun and the name of a feature called 'Bitwarden Send'. It should not be translated."
```

cscharf commented on src/locales/en/messages.json:
```diff
@@ -3635,6 +3638,26 @@
+    "message": "Send Options"
+  },
+  "sendOptionsPolicyDesc": {
+    "message": "Set options for creating and editing Sends."
```
```suggestion
    "message": "Set options for creating and editing Sends.",
    "description": "'Sends' is a plural noun and the name of a feature called 'Bitwarden Send'. It should not be translated."
```

cscharf reviewed (commented):
Just a few comments/feedback, would like the team to weigh in on the rest.

eliykat commented:
Note: build will be broken by unrelated jslib changes that got pulled in

addisonbeck commented on src/locales/en/messages.json:
```diff
@@ -3775,6 +3800,10 @@
       }
     }
   },
+  "viewSendHiddenEmailWarning": {
+    "message": "The Bitwarden user who created this Send has chosen to hide their email address. You should ensure you trust the source of this link before using or downloading its content.",
+    "description": "'Send' is a noun and the name of a feature called 'Bitwarden Send'. It should not be translated."
+  },
```
Good idea implementing this warning bit

addisonbeck commented on src/app/organizations/manage/policy-edit.component.ts:
```diff
@@ -60,6 +60,9 @@ export class PolicyEditComponent implements OnInit {
     passGenCapitalize?: boolean;
     passGenIncludeNumber?: boolean;
 
+    // Send options
+    sendDisableHideEmail?: boolean;
+
```
@cscharf mentioned to remove the line break between this comment and the options, but I see that that is not the style we are using for the other sections here (though I would say it should be). Can we sync these up and remove the other extra lines while in the area?

eliykat replied:
I agree that removing the whitespace is more consistent with our code style. Done

addisonbeck reviewed (commented):
Same as desktop, I went ahead and resolved the build issues here because they were insignificant. Other than that just a quick formatting comment.

eliykat commented:
The ubuntu build is using `git submodule update --remote` which is pulling in the wrong version of `jslib` - please ignore.

vachan-maker commented:
Sorry to interfere,
But will there be an option to report a malicious send? Or what about a send which contains some copyrighted material?
A Report Abuse button would be a nice.

addisonbeck commented:
@vachan-maker I think there is certainly a need to improve in that space, as right now there is not. We have been talking through various consequence paths for malicious use of Send, but I don't think a "Report Abuse" button of some kind has come up yet in the conversation. I think that would be a great feature request for our [community forums](https://community.bitwarden.com/) if you are interested in posting it there.

addisonbeck reviewed (dismissed):
I updated jslib & did a rebase to fix merge conflicts. Let me know if you feel some kind of way about me pushing to your PRs @eliykat, it just seems like bad asynchronous work to let little stuff like that sit until until you're on and nobody else is around to review anything 😅
